### ISSUE TITLE
Getting started on AWS EMR setup not working.

### ISSUE THREAD

**Issue by h4gen** [jupyter/enterprise_gateway#476]
Hello everyone!

First let me thank you for your work! I think what you develop here is what a lot of people - including me - are hoping and waiting for!

I tried to replicate your Getting started on an AWS EMR cluster with a spark configuration.
I think I did everything like explained but it wont really work.

== CLUSTER MASTER NODE ==

1. There seems to be no `SPARK_HOME` defined on the master node. As all applications seem to be installed in `/usr/lib/` on EMR I pointed `SPARK_HOME` to `/user/lib/spark`.

2. I set `EG_YARN_ENDPOINT` to the endpoint of my master, which seems to be running as I get following answer from the address:

```
<clusterInfo>
<id>1540680500465</id>
<startedOn>1540680500465</startedOn>
<state>STARTED</state>
<haState>ACTIVE</haState>
<rmStateStoreName>
org.apache.hadoop.yarn.server.resourcemanager.recovery.NullRMStateStore
</rmStateStoreName>
<resourceManagerVersion>2.8.4-amzn-1</resourceManagerVersion>
<resourceManagerBuildVersion>
2.8.4-amzn-1 from 29dbe82f1ecfd8384b89f1a32e5e2d9e5dfd7724 by ec2-user source checksum ec9c07392c6e193ae33f9c457530a6
</resourceManagerBuildVersion>
<resourceManagerVersionBuiltOn>2018-08-15T00:22Z</resourceManagerVersionBuiltOn>
<hadoopVersion>2.8.4-amzn-1</hadoopVersion>
<hadoopBuildVersion>
2.8.4-amzn-1 from 29dbe82f1ecfd8384b89f1a32e5e2d9e5dfd7724 by ec2-user source checksum 4057d63d697b1379eaf93c5ff6a9b96
</hadoopBuildVersion>
<hadoopVersionBuiltOn>2018-08-15T00:19Z</hadoopVersionBuiltOn>
<haZooKeeperConnectionState>
Could not find leader elector. Verify both HA and automatic failover are enabled.
</haZooKeeperConnectionState>
</clusterInfo>
```

3. I setup the new kernel which is then located in `/home/hadoop/anaconda3/share/jupyter/kernels/spark_python_yarn_cluster` as I use Anaconda. 
The `kernel.json` looks like this:

```
{
  "language": "python",
  "display_name": "Spark - Python (YARN Cluster Mode)",
  "process_proxy": {
    "class_name": "enterprise_gateway.services.processproxies.yarn.YarnClusterProcessProxy"
  },
  "env": {
    "SPARK_HOME": "/usr/hdp/current/spark2-client",
    "PYSPARK_PYTHON": "/opt/anaconda2/bin/python",
    "PYTHONPATH": "${HOME}/.local/lib/python2.7/site-packages:/usr/hdp/current/spark2-client/python:/usr/hdp/current/spark2-client/python/lib/py4j-0.10.4-src.zip",
    "SPARK_OPTS": "--master yarn --deploy-mode cluster --name ${KERNEL_ID:-ERROR__NO__KERNEL_ID} --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.appMasterEnv.PYTHONUSERBASE=/home/yarn/.local --conf spark.yarn.appMasterEnv.PYTHONPATH=${HOME}/.local/lib/python2.7/site-packages:/usr/hdp/current/spark2-client/python:/usr/hdp/current/spark2-client/python/lib/py4j-0.10.4-src.zip --conf spark.yarn.appMasterEnv.PATH=/opt/anaconda2/bin:$PATH",
    "LAUNCH_OPTS": ""
  },
  "argv": [
    "/usr/local/share/jupyter/kernels/spark_python_yarn_cluster/bin/run.sh",
    "{connection_file}",
    "--RemoteProcessProxy.response-address",
    "{response_address}",
    "--RemoteProcessProxy.port-range",
    "{port_range}",
    "--RemoteProcessProxy.spark-context-initialization-mode",
    "lazy"
  ]
}
```

So far everything seems to be okay as far as I can tell but when I

4. start the gateway with
```
jupyter enterprisegateway --ip=0.0.0.0 --port_retries=0
```

and check the gateway on YARN_MASTER_IP:8888, I get the following response:

```
{
  "reason": "Not Found",
  "message": ""
}
```

== LOCAL LAPTOP ==

5. Nevertheless I started the provided container on my laptop, which results in an totaly empty jupyterlab with no applications in the starter.

What am I doing wrong?
Thanks a lot!

Hagen

**Comment by kevin-bates** [comment]
@h4gen - thank you for your issue. AWS is something we've wanted to take for a spin but haven't had a chance.  So this is well-timed and of high interest to us.

While there's a possibility we need to create another [process proxy subclass](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/system-architecture.html#process-proxy) for this, let's see how far we can get prior to doing that.

One thing to keep in mind is that the kernelspec configurations we provide should be considered examples.  The yarn-based kernelspecs should work out of the box against HDP systems provided the installation directories are the same (`/usr/hdp`).  This is probably the area we should focus on first.

I see that you set `SPARK_HOME` to `/user/lib/spark` (I think you meant `/usr/lib/spark`) which is fine.  However, that same value should be specified in the kernelspecs `env:` stanza.  Any existing values prefixed with `/usr/hdp` will need to be analyzed and updated with appropriate paths relative to your AWS environment and the value's target file or directory.  Generally speaking, *all* path values in the `kernel.json` and its corresponding `run.sh` file should be validated.

If you haven't already, you should probably validate your Spark and YARN environments are in working order outside of Enterprise Gateway.  There should be an example for running one of the spark examples in cluster mode.  Please ensure those components are working independently.

In these early troubleshooting stages of an installation, I highly recommend you add `--log-level=DEBUG` to your command line and redirect `stdout` and `stderr` to a file.  We commonly invoke something like the following...
```bash
jupyter enterprisegateway --ip=0.0.0.0 --port_retries=0 --log-level=DEBUG > eg.log 2>&1 &
```
This places the EG server into the background and the log can then be tailed using `tail -f eg.log`.

>and check the gateway on YARN_MASTER_IP:8888, I get the following response:

How are you going about checking the gateway?  I would recommend issuing the following curl command to retrieve the set of configured kernelspecs:
```
curl -X GET http://<gateway-server>:<gateway-port>/api/kernelspecs
```
You should also see an entry in the EG log for each kernelspec found.

Once you have independently confirmed Spark and YARN and can retrieve the set of kernelspecs, then you're ready to issue a kernel creation request against the EG server.

Since you want to use Jupyterlab, you'll need to install the [NB2KG extension](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/getting-started.html#connecting-a-notebook-to-enterprise-gateway) if you haven't already.  Another option would be to use our [nb2kg docker image](https://hub.docker.com/r/elyra/nb2kg/) since it supports both notebook and lab clients.  Please note that I think the instructions to invoke lab may have changed recently.  @lresende, can you please confirm the [instructions on DockerHub](https://hub.docker.com/r/elyra/nb2kg/)?

Once the NB2KG extension is in place, then all kernelspecs and kernel management operations will be directed to your Enterprise Gateway configuration.  At that point, we can really get down to business with troubleshooting AWS-specific issues to determine whether an additional process proxy implementation is necessary. :smiley:

**Comment by lresende** [comment]
@kevin-bates , I believe you have given a pretty good summary of the steps @h4gen should take. As for the docker usage, only remove work from the home folder and all should be good (I will update the readme soon).

**Comment by h4gen** [comment]
@kevin-bates  Thanks for the answer and sorry for my late answer. I will give it a try on the weekend and get back to you!

**Comment by h4gen** [comment]
So I gave it another try today. Again `SPARK_HOME=/usr/lib/spark`. I changed all the entries in the kernel json accordingly. The user on an AWS EMR cluster is `hadoop` and I used the current anaconda3 version. This is the kernel.json after I changed all paths:  

```
{
  "language": "python",
  "display_name": "Spark - Python (YARN Cluster Mode)",
  "process_proxy": {
    "class_name": "enterprise_gateway.services.processproxies.yarn.YarnClusterProcessProxy"
  },
  "env": {
    "SPARK_HOME": "/usr/lib/spark",
    "PYSPARK_PYTHON": "/home/hadoop/anaconda3/bin/python",
    "PYTHONPATH": "${HOME}/.local/lib/python2.7/site-packages:/usr/lib/spark/python:/usr/lib/spark/python/lib/py4j-0.10.4-src.zip",
    "SPARK_OPTS": "--master yarn --deploy-mode cluster --name ${KERNEL_ID:-ERROR__NO__KERNEL_ID} --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.appMasterEnv.PYTHONUSERBASE=/home/hadoop/.local --conf spark.yarn.appMast$
    "LAUNCH_OPTS": ""
  },
  "argv": [
    "/home/hadoop/anaconda3/share/jupyter/kernels/spark_python_yarn_cluster/bin/run.sh",
    "{connection_file}",
    "--RemoteProcessProxy.response-address",
    "{response_address}",
    "--RemoteProcessProxy.port-range",
    "{port_range}",
    "--RemoteProcessProxy.spark-context-initialization-mode",
    "lazy"
  ]
}
```
As far as I can tell the `run.sh` had not to be changed at all as long as all variables are correctly set.
This is the result I get from `curl -X GET http://<gateway-server>:<gateway-port>/api/kernelspecs
`:

```
{
  "default": "python3",
  "kernelspecs": {
    "spark_python_yarn_cluster": {
      "name": "spark_python_yarn_cluster",
      "spec": {
        "argv": [
          "/home/hadoop/anaconda3/share/jupyter/kernels/spark_python_yarn_cluster/bin/run.sh",
          "{connection_file}",
          "--RemoteProcessProxy.response-address",
          "{response_address}",
          "--RemoteProcessProxy.port-range",
          "{port_range}",
          "--RemoteProcessProxy.spark-context-initialization-mode",
          "lazy"
        ],
        "env": {
          "SPARK_HOME": "/usr/lib/spark",
          "PYSPARK_PYTHON": "/home/hadoop/anaconda3/bin/python",
          "PYTHONPATH": "${HOME}/.local/lib/python2.7/site-packages:/usr/lib/spark/python:/usr/lib/spark/python/lib/py4j-0.10.4-src.zip",
          "SPARK_OPTS": "--master yarn --deploy-mode cluster --name ${KERNEL_ID:-ERROR__NO__KERNEL_ID} --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.appMasterEnv.PYTHONUSERBASE=/home/hadoop/.local --conf spark.yarn.appMasterEnv.PYTHONPATH=${HOME}/.local/lib/python2.7/site-packages:/usr/lib/spark/python:/usr/lib/spark/python/lib/py4j-0.10.4-src.zip --conf spark.yarn.appMasterEnv.PATH=/home/hadoop/anaconda3/bin:$PATH",
          "LAUNCH_OPTS": ""
        },
        "display_name": "Spark - Python (YARN Cluster Mode)",
        "language": "python",
        "interrupt_mode": "signal",
        "metadata": {
          
        },
        "process_proxy": {
          "class_name": "enterprise_gateway.services.processproxies.yarn.YarnClusterProcessProxy",
          "config": {
            
          }
        }
      },
      "resources": {
        
      }
    },
    "python3": {
      "name": "python3",
      "spec": {
        "argv": [
          "/home/hadoop/anaconda3/bin/python",
          "-m",
          "ipykernel_launcher",
          "-f",
          "{connection_file}"
        ],
        "env": {
          
        },
        "display_name": "Python 3",
        "language": "python",
        "interrupt_mode": "signal",
        "metadata": {
          
        },
        "process_proxy": {
          "class_name": "enterprise_gateway.services.processproxies.processproxy.LocalProcessProxy",
          "config": {
            
          }
        }
      },
      "resources": {
        "logo-64x64": "/kernelspecs/python3/logo-64x64.png",
        "logo-32x32": "/kernelspecs/python3/logo-32x32.png"
      }
    }
  }
}
```

Again I launched the container with:

```
docker run -t --rm \
  -e KG_URL='http://<master ip>:8888' \
  -e KG_HTTP_USER=guest \ 
  -e KG_HTTP_PASS=guest-password \ 
  -p 8888:8888 \
  -e VALIDATE_KG_CERT='no' \
  -e LOG_LEVEL=DEBUG \
  -e KG_REQUEST_TIMEOUT=40 \
  -e KG_CONNECT_TIMEOUT=40 \
  -v ${HOME}/notebooks/:/tmp/notebooks \
  -w /tmp/notebooks \
  elyra/nb2kg lab
```

Resulting in an empty jupyterlab :(

![image](https://user-images.githubusercontent.com/12556763/47955204-8246b700-df6a-11e8-8850-f3c9a799f058.png)

Spark and Yarn should be running correctly. I can connect to the Zeppelin notebook and execute the examples.

**Comment by lresende** [comment]
A few questions below:

- What's the URL you use to access the Notebook UI? 
- As you are getting a proper response from eg/api/kernelspecs it seems that the issue is more on the lab setup. 
- Any difference if you try Jupyter Notebook versus Lab which should be mainly `/` versus `/lab`?


Also, could you try:

```
docker run -t --rm \
  -e KG_URL='http://<master ip>:8888' \
  -e KG_HTTP_USER=guest \ 
  -e KG_HTTP_PASS=guest-password \ 
  -p 8888:8888 \
  -e VALIDATE_KG_CERT='no' \
  -e LOG_LEVEL=DEBUG \
  -e KG_REQUEST_TIMEOUT=40 \
  -e KG_CONNECT_TIMEOUT=40 \
  -v ${HOME}/notebooks/:/tmp/notebooks \
  -w /tmp/notebooks \
  elyra/nb2kg:2.0.0.dev0
``` 
And then access it using a url similar to the one below for notebook:
`http://localhost:8080/`
or for lab
`http://localhost:8888/lab`

 I will try to set up an AWS env over the weekend and see if I see similar issues.

**Comment by h4gen** [comment]
> A few questions below:
> 
> * What's the URL you use to access the Notebook UI?
I use localhost:8888
> * As you are getting a proper response from eg/api/kernelspecs it seems that the issue is more on the lab setup.
What about if I get a response, but maybe still did something wrong during configuration?
> * Any difference if you try Jupyter Notebook versus Lab which should be mainly `/` versus `/lab`?
Nope, both don't work :(

**Comment by kevin-bates** [comment]
A couple items worth pointing out to make sure we've "connected the dots".

1. The `KG_URL` used in the docker command needs to match the address used in your curl command.  So be sure to use `-e KG_URL='http://<gateway-server>:<gateway-port>'`.
2. I would add a container name parameter to the docker run command, this makes it easier for the communicating troubleshooting.  Let's use a name of `nb_to_emr` for this exercise.
3. Looks like Notebook doesn't allow a way to enable DEBUG logging via an env variable, so I hope we won't need that.
4. The startup script assumes any existing notebooks will be located in $HOME/notebooks. Change that portion of the -v parameter if you'd like your notebooks in another location.

Here's the result of those items:
```
docker run -t --rm \
  --name nb_to_emr \
  -e KG_URL='http://<gateway-server>:<gateway-port>' \
  -e KG_HTTP_USER=guest \
  -e KG_HTTP_PASS=guest-password \
  -p 8888:8888 \
  -e VALIDATE_KG_CERT='no' \
  -e KG_REQUEST_TIMEOUT=40 \
  -e KG_CONNECT_TIMEOUT=40 \
  -v ${HOME}/notebooks/:/tmp/notebooks \
  -w /tmp/notebooks \
  elyra/nb2kg:2.0.0.dev0
```
Once the container is launched and assuming you see similar results (i.e., nothing), please provide the output of:
```
docker logs nb_to_emr
```

> What about if I get a response, but maybe still did something wrong during configuration?

As Luciano points out, given that you can retrieve the kernelspecs from the gateway indicates that gateway is picking up the right information and exposing it at `http://<gateway-server>:<gatewayport>`.

Once we get your notebook client seeing those same kernelspecs, then that's when the fun actually begins.  There will likely be some iterations getting the kernels running in EMR.  Since you also have the default `python3` kernel available, I would recommend confirming its operation first.  This will validate that the rest of EG is behaving correctly.  Then we can move into the launching a kernel in the EMR cluster.

**Comment by h4gen** [comment]
@kevin-bates Sorry, I was busy with getting my JupyterHub running. Will give it a try the next days. But I switched to kubernetes for my general setup, so I will check it and provide my insights but then I will try to integrate the enterprise gateway in a JupyterHub on Kubernetes environment.

**Comment by kevin-bates** [comment]
Awesome - thanks!  Btw, we're preparing for our 2.0 release - featuring Kubernetes support, so any feedback would highly appreciated.  You may also find this [blog post](https://blog.jupyter.org/on-demand-notebooks-with-jupyterhub-jupyter-enterprise-gateway-and-kubernetes-e8e423695cbf) helpful.

**Comment by kevin-bates** [comment]
Closing this specific issue.  Others can refer to @lresende's previous comment for EMR config info and re-open or create a new issue if necessary.  Thanks.
